### DISCUSSION TITLE
Docker run -v inside container

### DISCUSSION THREAD

**Discussion by mattvb91** [community/community#26360] [Actions]
<p>Hey everyone,</p>
<p>im trying to do a “docker run -v” with a volume inside a container but the mounted directory seems to come back empty.</p>
<pre><code class="lang-auto">name: CI

on:
  push:
  pull_request:

jobs:
  build:
    runs-on: ubuntu-latest
    container: 
      image: lambci/lambda:build-nodejs12.x

    steps:
      - uses: actions/checkout@v2

      # Runs a set of commands using the runners shell
      - name: Build
        run: |
          pwd
          ls -lah $GITHUB_WORKSPACE/lambda/hooks
          docker run -v "$GITHUB_WORKSPACE/lambda/hooks":/var/task bash ls -lah /var/task
</code></pre>
<p>The last ls -lah executed with the <code>bash</code> image just returns an empty directory whereas the first <code>ls -lah $GITHUB_WORKSPACE/lambda/hooks</code> shows all the files existing.</p>
<p>Is it possible to mount again inside a running container?</p>

**Comment by airtower-luna** [comment]
<p>For job containers the Docker control socket is mounted from the runner VM into the container, so the volume paths need to be paths as on the runner. In particular, I’ve encountered the workspace paths being different (also via volume mounts into the job container). From the snippet above I’m not sure how the paths would need to look to work, but that’s a likely reason.</p>

**Comment by mattvb91** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a> thanks for the suggestion. Any chance you have some links to more reading? Im afraid ive exhausted all my search terms.</p>
<p>The closest i’ve gotten to finding something is this: <a href="https://github.com/GoogleCloudPlatform/cloud-builders/issues/372#issuecomment-531311995" rel="noopener nofollow ugc">https://github.com/GoogleCloudPlatform/cloud-builders/issues/372#issuecomment-531311995</a></p>
<pre><code class="lang-auto">volumes:
  - name: 'vol1'
    path: '/persistent_volume'
</code></pre>
<p>But I think the -name path: syntax has been deprecated on github actions</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>That has never been Github Actions syntax as far as I know, the stuff in that repository is for Google Cloud Build, which looks a bit similar because it’s YAML, too.</p>
<p>The point I was trying to make is that your job runs in a container (the <code>lambci/lambda:build-nodejs12.x</code> one). The paths inside that container are not going to be the same as on the <code>ubuntu-latest</code> runner VM.</p>
<p><code>GITHUB_WORKSPACE</code> is an example I specifically had trouble with while uploading SARIF for the security tab. The <a href="https://github.com/github/codeql-action/blob/8d18e347a7c155d034848ceeb01415097f59fd07/upload-sarif/action.yml"><code>github/codeql-action/upload-sarif</code> action</a> needs to make any absolute paths in the file to be uploaded relative (so any issues found can be assigned to files in the repository). By default it assumes the base path for the repository is <code>$GITHUB_WORKSPACE</code>. That works nicely if you build <em>directly on the runner VM</em>. In a job container the workspace is mounted to <code>/__w/</code>, and the repository will be below that. I ended up <a href="https://github.com/airtower-luna/mod_gnutls/commit/6be4dd05fcd7dd52805bedef263e85e4be0b6747">using the container workdir instead</a> to fix the problem.</p>
<p>If you want to see what is mounted where in your job container, check the “Initialize containers” step near the start of your workflow. You’ll notice that the Docker command to create the job container includes this:</p>
<pre><code class="lang-nohighlight">-v "/var/run/docker.sock":"/var/run/docker.sock"
</code></pre>
<p>This means any Docker command in the container will talk to the Docker daemon on the runner VM. So any path for volume mounts will be interpreted from the runner VM, not inside your container.</p>

**Comment by mattvb91** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a> thank you!! Put me on the right path.</p>
<p>I managed to get working now after looking at the <code>-v</code> options and attempted to mount a new volume there:</p>
<pre><code class="lang-auto">    container: 
      image: lambci/lambda:build-nodejs12.x
      options: -v /volume_mount:/volume_mount
</code></pre>
<p>and then later after checking out the repo I</p>
<pre><code class="lang-auto">          cp -R /__w/my_dir /volume_mount
</code></pre>
<p>Now inside the next container I can finally see the contents. Still a bit rough on the edges but at least its a start.</p>
<p>Thanks again!</p>
